rust/gui-client/src-tauri/src/client/tunnel-wrapper/ipc.rs

```diff
@@ -0,0 +1,145 @@
+    tx: FramedWrite<OwnedWriteHalf, LengthDelimitedCodec>,
+}
+
+impl TunnelWrapper {
+    pub(crate) async fn disconnect(mut self) -> Result<()> {
+        self.send_msg(&IpcClientMsg::Disconnect)
+            .await
+            .context("Couldn't send Disconnect")?;
+        self.tx.close().await?;
+        self.recv_task.abort();
+        Ok(())
+    }
+
+    pub(crate) async fn reconnect(&mut self) -> Result<()> {
+        self.send_msg(&IpcClientMsg::Reconnect)
+            .await
+            .context("Couldn't send Reconnect")?;
+        Ok(())
+    }
+
+    /// Tell connlib about the system's default resolvers
+    ///
+    /// `dns` is passed as value because the in-proc impl needs that
+    pub(crate) async fn set_dns(&mut self, dns: Vec<IpAddr>) -> Result<()> {
+        self.send_msg(&IpcClientMsg::SetDns(dns))
+            .await
+            .context("Couldn't send SetDns")?;
+        Ok(())
+    }
+
+    async fn send_msg(&mut self, msg: &IpcClientMsg) -> Result<()> {
+        self.tx
+            .send(
+                serde_json::to_string(msg)
+                    .context("Couldn't encode IPC message as JSON")?
+                    .into(),
+            )
+            .await
+            .context("Couldn't send IPC message")?;
+        Ok(())
+    }
+}
+
+pub async fn connect(
+    api_url: &str,
+    token: SecretString,
+    callback_handler: CallbackHandler,
+    tokio_handle: tokio::runtime::Handle,
+) -> Result<TunnelWrapper> {
+    tracing::info!(pid = std::process::id(), "Connecting to IPC service...");
+    let stream = UnixStream::connect(sock_path())
+        .await
+        .context("Couldn't connect to UDS")?;
+    let (rx, tx) = stream.into_split();
+    let mut rx = FramedRead::new(rx, LengthDelimitedCodec::new());
+    let tx = FramedWrite::new(tx, LengthDelimitedCodec::new());
+
+    // TODO: Make sure this joins / drops somewhere
+    let recv_task = tokio_handle.spawn(async move {
+        while let Some(msg) = rx.next().await {
+            let msg = msg?;
+            let msg: IpcServerMsg = serde_json::from_slice(&msg)?;
+            match msg {
+                IpcServerMsg::Ok => {}
+                IpcServerMsg::OnDisconnect => callback_handler.on_disconnect(
+                    &connlib_client_shared::Error::Other("errors can't be serialized"),
+                ),
+                IpcServerMsg::OnUpdateResources(v) => callback_handler.on_update_resources(v),
+                IpcServerMsg::TunnelReady => callback_handler.on_tunnel_ready(),
+            }
+        }
+        Ok(())
+    });
+
+    let mut client = TunnelWrapper { recv_task, tx };
+    let token = token.expose_secret().clone();
+    client
+        .send_msg(&IpcClientMsg::Connect {
+            api_url: api_url.to_string(),
+            token,
+        })
+        .await
+        .context("Couldn't send Connect message")?;
+
+    Ok(client)
+}
+
+// Callbacks must all be non-blocking
+// TODO: DRY
```

Comment:
There are 4 of these `CallbackHandler` structs now, but they're all subtly different, so I didn't try to de-dupe them yet.

I considered trying to fit them all into a "wake-and-poll" pattern but this would probably make the code longer, not shorter, and it would make the IPC protocol a little more chatty. (e.g. instead of the service just sending new resources to the Client, it would send a wake, then the Client would make a request, then the service would respond. I think this is more robust if the system is CPU-starved but in practice it probably doesn't matter at all.)